File: modules/expo-sticker-runtime/android/src/test/java/expo/modules/stickerruntime/ClipTokenizerTest.kt
package expo.modules.stickerruntime

import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Test

class ClipTokenizerTest {
  @Test
  fun `encodes and pads pinned CLIP BPE fixture to 77 tokens`() {
    val tokenizer = ClipTokenizer.fromJson(
      """
        {
          "model": {
            "vocab": {
              "<|startoftext|>": 49406,
              "<|endoftext|>": 49407,
              "a</w>": 320,
              "cat</w>": 2368
            },
            "merges": [["a","t</w>"],["c","a"],["c","at</w>"]]
          }
        }
      """.trimIndent(),
    )

    val ids = tokenizer.encode("A CAT")

    assertEquals(77, ids.size)
    assertArrayEquals(longArrayOf(49406, 320, 2368, 49407), ids.copyOfRange(0, 4))
    assertEquals(49407, ids.last())
  }

  @Test
  fun `truncation always retains the end token`() {
    val tokenizer = ClipTokenizer.fromJson(
      """
        {"model":{"vocab":{"<|startoftext|>":49406,"<|endoftext|>":49407,"a</w>":320},"merges":[]}}
      """.trimIndent(),
    )

    val ids = tokenizer.encode(List(100) { "a" }.joinToString(" "))

    assertEquals(49406, ids.first())
    assertEquals(49407, ids.last())
    assertEquals(75, ids.count { it == 320L })
  }
}

File: src/services/generation/web/clipTokenizer.test.ts
import { describe, expect, test } from 'vitest';
import { ClipTokenizer } from './clipTokenizer';

const PINNED_TOKENIZER = JSON.stringify({
  model: {
    vocab: {
      '<|startoftext|>': 49406,
      '<|endoftext|>': 49407,
      'a</w>': 320,
      'cat</w>': 2368,
    },
    merges: [
      ['a', 't</w>'],
      ['c', 'a'],
      ['c', 'at</w>'],
    ],
  },
});

describe('ClipTokenizer web parity', () => {
  test('encodes and pads the Kotlin CLIP BPE fixture to 77 tokens', () => {
    const ids = ClipTokenizer.fromJson(PINNED_TOKENIZER).encode('A CAT');

    expect(ids).toHaveLength(77);
    expect(ids.slice(0, 4)).toEqual([49406n, 320n, 2368n, 49407n]);
    expect(ids.at(-1)).toBe(49407n);
  });

  test('truncates content while retaining the terminal token', () => {
    const ids = ClipTokenizer.fromJson(PINNED_TOKENIZER).encode(
      Array.from({ length: 100 }, () => 'a').join(' '),
    );

    expect(ids[0]).toBe(49406n);
    expect(ids.at(-1)).toBe(49407n);
    expect(ids.filter((id) => id === 320n)).toHaveLength(75);
  });

  test('rejects a tokenizer with incompatible special-token IDs', () => {
    const invalid = JSON.stringify({
      model: {
        vocab: { '<|startoftext|>': 1, '<|endoftext|>': 2 },
        merges: [],
      },
    });

    expect(() => ClipTokenizer.fromJson(invalid)).toThrow('special token');
  });
});
